File: go/queue.go
package main

import (
	"fmt"
	"sort"
	"strconv"
)

// Структура для хранения информации о посетителе - время и номер
type Visitor struct {
	time  int
	coupon string
}

// Глобальная переменная для выдачи уникальных талончиков
var counter = 0

// Функция для выдачи уникального талончика
func Coupon() string {
	counter++
	return "T" + strconv.Itoa(counter-1)
}

// Функция для добавления посетителя в очередь
func inEnqueue(queue *[]Visitor, time int) {
	visitor := Visitor{time: time, coupon: Coupon()}
	*queue = append(*queue, visitor)
	fmt.Println(visitor.coupon)
}

// Функция для распределения посетителей по окнам
func okno(queue *[]Visitor, countWindows int) {
	windows := make([][]Visitor, countWindows) // Посетителей в каждом окне
	windowEndTimes := make([]int, countWindows) // Время окончания обслуживания в каждом окне

	// Сортируем посетителей по убыванию времени обслуживания
	sort.Slice(*queue, func(i, j int) bool {
		return (*queue)[i].time > (*queue)[j].time
	})

	// Распределяем посетителей по окнам
	for _, visitor := range *queue {
		// Найти окно с минимальным временем окончания обслуживания - индекс
		minIndex := 0
		for i := 1; i < countWindows; i++ {
			if windowEndTimes[i] < windowEndTimes[minIndex] {
				minIndex = i
			}
		}
		windows[minIndex] = append(windows[minIndex], visitor)
		windowEndTimes[minIndex] += visitor.time
	}

	// Вывод информации о распределении посетителей по окнам
	for i := 0; i < countWindows; i++ {
		fmt.Printf("Окно %d (%d минут):", i+1, windowEndTimes[i])
		for _, visitor := range windows[i] {
			fmt.Printf(" %s", visitor.coupon)
		}
		fmt.Println()
	}
}

func main() {
	var countWindows int
	fmt.Print("Кол-во окон: ")
	fmt.Scan(&countWindows)

	var queue []Visitor
	var command string
	var time int

	for {
		fmt.Scan(&command)
		if command == "ENQUEUE" {
			fmt.Scan(&time)
			inEnqueue(&queue, time)
		} else if command == "DISTRIBUTE" {
			okno(&queue, countWindows)
			break
		}
	}
}

File: queue.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <sstream>
#include <algorithm>
#include <iomanip>

using namespace std;

// Структура для хранения информации о посетителе - время и номер
struct Visitor {
    int time;
    string coupon;
};

// выдача талончика
string Coupon() {
    static int counter = 0;
    ostringstream oss; 
    oss << "T" << counter++;
    return oss.str();
}

// добавление посетителя в очередь - время и номер - заносим в вектор
void in_enqueue(queue<Visitor>& queue, int time) {
    Visitor visitor;
    visitor.time = time;  // время
    visitor.coupon = Coupon();  // талончик
    queue.push(visitor);
    cout << visitor.coupon << endl;
}

// распределения посетителей по окнам
void okno(queue<Visitor>& queue, int count_windows) {
    vector<vector<Visitor>> windows(count_windows);  // посетителей в каждом окне
    vector<int> windowEndTimes(count_windows, 0);  // времени окончания обслуживания в каждом окне

    vector<Visitor> visitors;  // все посетители
    while (!queue.empty()) {
        visitors.push_back(queue.front());
        queue.pop();
    }

    // Сортируем посетителей по убыванию времени обслуживания
    sort(visitors.begin(), visitors.end(), [](const Visitor& a, const Visitor& b) {
        return a.time > b.time;
    });

    // Распределяем посетителей по окнам
    for (const auto& visitor : visitors) {
        // Найти окно с минимальным временем окончания обслуживания - индекс
        int minindex = min_element(windowEndTimes.begin(), windowEndTimes.end()) - windowEndTimes.begin();
        windows[minindex].push_back(visitor);

        // время для этого окна
        windowEndTimes[minindex] += visitor.time;
    }

    for (int i = 0; i < count_windows; ++i) {
        cout << "Окно " << i + 1 << " (" << windowEndTimes[i] << " минут):";
        for (const auto& visitor : windows[i]) {
            cout << " " << visitor.coupon;
        }
        cout << endl;
    }
}

int main() {
    int count_windows;  // кол-во окон
    cout << "Кол-во окон: ";
    cin >> count_windows;

    queue<Visitor> queue;
    string command;
    int time;

    while (true) {
        cin >> command;
        if (command == "ENQUEUE") {
            cin >> time;
            in_enqueue(queue, time);
        } else if (command == "DISTRIBUTE") {
            okno(queue, count_windows);
            break;
        }
    }

    return 0;
}
